Add the ability to specify unwanted pokemon instead of wanted pokemon

everix1992 opened the pull request:
This pull request includes a
- [ ] Bug fix
- [X] New feature
- [ ] Translation

The following changes were made
- Add the ability to specify a list of pokemon to NOT be notified for in the config file. This will only be used when this list is present AND the list of pokemon to notify for (the "notify" option) is not present. If the "notify" list is present, this new list will not be used. Here's a sample config using the new "do_not_notify" option:

{
  "auth_service": "google",
  "username": "your_email_here",
  "password": "your_password_here",
  "step_limit": 5,
  "location": "1 White House Lane",
  "do_not_notify": "rattata,raticate,pidgey,pidgeotto,pidgeot,weedle,kakuna,caterpie,metapod,zubat,golbat,venonat,oddish,magikarp,spearow,paras,ekans,goldeen,doduo,pinsir,poliwag",
  "pushbullet": "o.m5fgAZaBFiZ0gHrNg21ZLFOKZStckql"
}

everix1992 committed:
Add the ability to specify unwanted pokemon instead of wanted pokemon

jxmorris12 commented:
still a lot of pushes. i'd like to offer both options

everix1992 commented:
This should still allow both options. It basically has three states:
1. If the "notify" and "do_not_notify" options are not present, you will be notified of all pokemon.
2. If the "notify" list is present, you will only be notified of pokemon in that list.
3. If the "notify" list is not present, but the "do_not_notify" list is present, you will be notified of all pokemon except the ones in the "do_not_notify" list.

everix1992 committed:
Add default typing for all arguments and convert based on that. This prevents some conversion errors when specifying the host

karlhe commented:
I've noticed that `main.py` already supports and handles both `ignore` and `only` in the options. So even the `notify` list in `notifier.py` is already duplicated functionality?

everix1992 commented:
@karlhe They are sort of different behaviors. Ignore and Only completely filter them from the map, whereas with "notify" and "do_not_notify" they are only filtered from notifications, but not the map.

jxmorris12 commented:
yeah exactly @everix1992

jxmorris12 commented:
I like this! can you also add an update to the `README`?

everix1992 commented:
@jxmorris12  Absolutely. I probably can't do it until later tonight as I'm out right now, but I will when I get home.

jxmorris12 commented:
awesome

everix1992 committed:
Remove an extra print line in the main arguments parser AND update the Readme for the new 'do_not_notify' option.

everix1992 commented:
@jxmorris12 I've now updated the `README` per your request. Note that some unrelated changes are also included in my branch (commit 423622a825529817aa7404822522d4b469b09ccf). These were made to fix some problems I had with specifying a different hostname in the config file. I just added a new list that defines the types of each argument and converts the argument value to that type if possible. Honestly, it would probably be best to use the python argument library, but I wasn't familiar enough with it to try doing it myself.

jxmorris12 commented on main.py:
```diff
@@ -445,19 +446,65 @@ def get_args():
         for key in data:
-            default_args[key] = str(data[key])
+            if key not in default_args_type:
+                warnings.warn( 'Config Item ' + key + 'Does Not Have a Default Type' )
```
cool

jxmorris12 commented:
this is really cool, great work
